### DISCUSSION TITLE
`Y` in block-visual mode

### DISCUSSION THREAD

**Discussion by simonmandlik** [neovim/neovim#29414] [Ideas]
We have had `Y` mapped to `y$` in normal mode for quite some now.
In character-visual and line-visual `Y` yanks whole lines.

But in block-visual mode `y` and `Y` are synonyms.

Would it make sense to make `Y` in block-visual mode yank whole lines (or until EOLs) as well?

**Comment by justinmk** [comment]
why?

**Reply by simonmandlik** [reply]
When I was studying about operators in visual mode, I conceptually split them into three groups:

1. operating on the exact selection (like `r`, `x`, `d`, `c`, `y`, `u`, `U`, `~`, ...)
2. always expanding to whole lines (like `S`, `R`, `X`, `J`)
3. expanding to whole lines, in blockwise visual mode expanding to ends of lines (like `D`, `C`, in line with their normal mode behavior)

In [docs](https://neovim.io/doc/user/visual.html) the categorization is done slightly differently (`C` is said to be operating on the block only, but then applying `C` on block changes to EOLs).

`Y` is in this taxonomy quite special and doesn't fit into any of the three groups, in blockwise visual mode it copies the whole block, same as `y` and other operators in the first group, in charwise it expands to whole lines like operators in the second group.

Since in normal mode, we have uppercase versions of `y`, `c` and `d` operating until EOLs, conceptually it would make more sense to me to have `Y` behave more like `C` and `D`, and  operate until EOLs. Similarly to how `Y` shortcuts `y$` in normal mode, `$y` would reduce to `Y` in blockwise visual mode. And the original functionality of `Y` would still remain accessible with `y`.

`Y` even in normal mode is IMO quite niche usecase, and this special behavior of `Y` in blockwise visual mode would be even less common. But it still sometimes catches me by surprise since I got used to `Y` in normal mode.

**Reply by justinmk** [reply]
> `Y` even in normal mode is IMO quite niche usecase

Nonsense.

**Reply by simonmandlik** [reply]
> Nonsense.

Would it make sense to have something similar in visual mode then?

**Comment by dundargoc** [comment]
I'm a bit confused, it seems you're both advocating for `Y` to copy entire line in

> Would it make sense to make Y in block-visual mode yank whole lines as well?

but also that is should only copy until end of line in

> conceptually it would make more sense to me to have `Y` behave more like `C` and `D`, and operate until EOLs.

Maybe I misunderstood what you meant.

In any case, I think it makes sense to have `Y` behave as `C` and `D` in blockwise visual mode for consistency and better mental model.

**Reply by simonmandlik** [reply]
@dundargoc thanks for the reply and sorry for the confusion!

What I meant is to ask why `y` and `Y` do not behave differently in block-visual mode, which seems like a missed opportunity to me (I tried explaining it further in [this comment](https://github.com/neovim/neovim/discussions/29414#discussioncomment-9837888)).

Both options — making `Y` copy whole lines, or making `Y` copy until EOLs — seem fine to me. Personally, I would slightly prefer the latter.

I have edited the wording of the original question to make it more clear.

**Reply by dundargoc** [reply]
> What I meant is to ask why y and Y do not behave differently in block-visual mode, which seems like a missed opportunity to me (I tried explaining it further in this comment).

Likely for historical reasons since `Y` has always been special. I don't think anyone would be against if someone sent a PR to make `Y` behave like `D` and `C`.
